### PR TITLE
drivers: Backlight PWM driver

### PR DESCRIPTION
**Pull request by MrMarteng**
Adding backlight driver api to enable possibility to control display backlight commonly.
This PR includes PWM version of a backlight driver.

I would like to bring up the discussion about the possibilities for a flexible way to control backlight. There currently exists no possibility to automatically initialise backlight on boot up. There also is no common way to control backlights (e.g. API).  

@faxe1008 brought up the idea to integrate backlight control into display drivers (#70663). I am emphasising creation of a dedicated backlight API just like it is used on linux. Imo a simple and unique backlight API offers best flexibility. This PR adds a PWM backlight driver and a general backlight API. Further backlight implementations can follow (e.g. via buck, gpio,... ). The overhead is minimal and it is an optional feature.

### CODE REVIEW

File: dts/bindings/display/backlight-pwm.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2024, Maritn Kiepfer
+# SPDX-License-Identifier: Apache-2.0
+
+description: Backlight PWM node
+
+compatible: "backlight-pwm"
```
Comment (jilaypandya) [inline dts/bindings/display/backlight-pwm.yaml:6]: probably use zephyr as vendor? `zephyr, backlight-pwm`

File: drivers/display/CMakeLists.txt [hunk 2]
```diff
@@ -31,6 +31,7 @@ zephyr_library_sources_ifdef(CONFIG_GC9X01X		display_gc9x01x.c)
 zephyr_library_sources_ifdef(CONFIG_LED_STRIP_MATRIX	display_led_strip_matrix.c)
 zephyr_library_sources_ifdef(CONFIG_DISPLAY_RENESAS_LCDC display_renesas_lcdc.c)
 zephyr_library_sources_ifdef(CONFIG_NT35510		display_nt35510.c)
+zephyr_library_sources_ifdef(CONFIG_BACKLIGHT_PWM		backlight_pwm.c)
```
Comment (faxe1008) [inline drivers/display/CMakeLists.txt:34]: Since its a full blown api I'd go for a seperate directory.

File: drivers/display/Kconfig.backlight_pwm [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+config BACKLIGHT_PWM
+	bool "Backlight PWM driver"
+	default y
+	depends on DT_HAS_BACKLIGHT_PWM_ENABLED
+	help
+	  Enable Backlight PWM driver.
+
+config BACKLIGHT_PWM_INIT_PRIORITY
```
Comment (faxe1008) [inline drivers/display/Kconfig.backlight_pwm:13]: This is usually a per driver class kconfig

File: dts/bindings/display/backlight.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    description: Initial value of the Backlight.
+
+  disable-init-on-boot:
+    type: boolean
+    description: |
+      Disable the initial value during bootup of the device. By default the
+      init-level is applied on boot. This option can prevent from this default
+      behaviour. Backlight needs to be enabled manually later in this case.
+
+  label:
```
Comment (jilaypandya) [inline dts/bindings/display/backlight.yaml:19]: label is already included in base.yaml. Is there a need to redefine it?

File: include/zephyr/display/backlight.h [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public Backlight driver APIs
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_BACKLIGHT_H_
```
Comment (jilaypandya) [inline include/zephyr/display/backlight.h:12]: `ZEPHYR_INCLUDE_DISPLAY_BACKLIGHT_H_` would be correct namespace

File: dts/bindings/display/backlight-pwm.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+include: [base.yaml, backlight.yaml]
```
Comment (faxe1008) [inline dts/bindings/display/backlight-pwm.yaml:8]: I think the backlight.yaml could heavily borrow the properties from linux
https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/backlight/common.yaml

This also allows for things like stepped brightness levels.

**Comment by faxe1008** [comment]
@MrMarteng I did some work on this way back 
https://github.com/faxe1008/zephyr/commits/backlight-driver/ but I postponed getting it in since I wanted to have the PR by @pdgendt (the one about giving drivers a class and placing them in dedicated linker sections) to get the full benefit of the backlight driver class. This way the backlight would be called automatically by the display driver code. I felt like this would be needed to justify the overhead of another API which is essentially a rewamp of the LED API, but maybe I am wrong on this. Conceptually the API makes sense since there are also dedicated IC for backlight control. So thanks for picking it up, if you want to feel free to take stuff from my PoC.

File: include/zephyr/display/backlight.h [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+/**
+ * @brief Turn on an Backlight
+ *
+ * This routine turns on Backlight. It will restore the previous set value
+ * or the initial value automatically.
+ *
+ * @param dev Backlight device
+ * @return 0 on success, negative on error
```
Comment (jilaypandya) [inline include/zephyr/display/backlight.h:116]: list all the errors, which the api would be propagated.  Refer to: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/can.h.

File: include/zephyr/display/backlight.h [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	if (api->set_brightness == NULL) {
+		return -ENOSYS;
+	}
+	return api->set_brightness(dev, value);
+}
+
+/**
+ * @brief Turn on an Backlight
+ *
+ * This routine turns on Backlight. It will restore the previous set value
```
Comment (jilaypandya) [inline include/zephyr/display/backlight.h:112]: @details

File: include/zephyr/display/backlight.h [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+#include <zephyr/device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Backlight information structure
+ *
+ * This structure gathers useful information about Backlight controller.
```
Comment (jilaypandya) [inline include/zephyr/display/backlight.h:34]: `useful information` leaves too much room for interpretation :-) Can it be more precise as to what exactly could be the use-cases of this struct

**Comment by MrMarteng** [comment]
> @MrMarteng I did some work on this way back https://github.com/faxe1008/zephyr/commits/backlight-driver/ but I postponed getting it in since I wanted to have the PR by @pdgendt (the one about giving drivers a class and placing them in dedicated linker sections) to get the full benefit of the backlight driver class. This way the backlight would be called automatically by the display driver code. I felt like this would be needed to justify the overhead of another API which is essentially a rewamp of the LED API, but maybe I am wrong on this. Conceptually the API makes sense since there are also dedicated IC for backlight control. So thanks for picking it up, if you want to feel free to take stuff from my PoC.

I didn't know this. Is there a PR for this change as well?
https://github.com/faxe1008/zephyr/commit/6f6a1d6fe86a83af718978a727e2f17618f319f2 looks very similar to this. I'll definitely will have a look into this.

**Comment by fabiobaltieri** [comment]
Don't think having a separate class is a good idea:
- do you see the need of decoupling a display from a backlight? I don't, then the backlight should be part of the display
- any display controller with integrated backlight control would now have to instantiate two drivers

Here's the situation: a display is going to have a backlight control, the implementation can be display controller specific, or a gpio, or a pwm.

How about making the backlight control api part of the display api itself, then a display can implement it directly and that would be the default handler, alternatively the devicetree can specify a pwm or a gpio and if that is present that is used instead. The logic can be at API level, this would require reworking the existing display drier to implement a common config structure to hold the subsystem level properties, this is already a thing in other subsystems (`input_kbd_matrix.h` is what I normally refer to but there are others as well, regulator, gpio...). Does it make sense?

**Comment by MrMarteng** [comment]
> Don't think having a separate class is a good idea:
> 
> * do you see the need of decoupling a display from a backlight? I don't, then the backlight should be part of the display
> * any display controller with integrated backlight control would now have to instantiate two drivers
> 
> Here's the situation: a display is going to have a backlight control, the implementation can be display controller specific, or a gpio, or a pwm.

In my opinion decoupling helps to avoid bloating up the display driver. Having a separate backlight driver reduces complexity. There might be various possibilities for controlling a backlight (buck, pwm, gpio,...). When integrating the backlight api into the display driver we would need to add all this functionality directly into the display driver, which would increase complexity in the display driver.

You brought up in my initial PR (https://github.com/zephyrproject-rtos/zephyr/pull/67012#discussion_r1444701046) display disturbances are often an issue when the initialization is not synchronised correctly. Ensuring this synchronisation in the display driver might not always work. When having separated drivers, more application specific solutions are possible.

**Comment by fabiobaltieri** [comment]
> When integrating the backlight api into the display driver we would need to add all this functionality directly into the display driver, which would increase complexity in the display driver.

The opposite of that, integration with backlight control supported by existing subsystem (gpio/pwm) should be at display api level so it only need to be implemented once, control using the driver specific capability can then live in the driver itself. Also the api is already there (`display_set_brightness_api`). I think I'm not explaining myself very well, may throw together a proof of concept myself so we can compare implementations maybe.

**Comment by MrMarteng** [comment]
> Also the api is already there (`display_set_brightness_api`). I think I'm not explaining myself very well, may throw together a proof of concept myself so we can compare implementations maybe. Waiting for @gmarull opinion as well btw.

Isn't #70663 the implementation you are referring to?

**Comment by fabiobaltieri** [comment]
> Isn't #70663 the implementation you are referring to?

Oh yeah it is, forgot about it.

**Comment by faxe1008** [comment]
Not quite, actually adding a `display-backlight-common` binding would be missing imo.

**Review by gmarull — CHANGES_REQUESTED** [review]
we need a proper backlight driver class.

**Comment by fabiobaltieri** [comment]
> we need a proper backlight driver class.

I'd argue that we'd be better off we something simpler and integrated with the display drivers like https://github.com/zephyrproject-rtos/zephyr/pull/70663, what do you think that backlight class would do in addition to the existing "set backlight" function?

**Comment by gmarull** [comment]
> > we need a proper backlight driver class.
> 
> I'd argue that we'd be better off we something simpler and integrated with the display drivers like #70663, what do you think that backlight class would do in addition to the existing "set backlight" function?

We have for example `auxdisplay` driver class, which also can benefit from the same generic backlight drivers. I also see the backlight API in both aux/disp APIs a workaround to our device model: only allows for a single device per DT node. So, if a display has built-in backlight control (like e.g. some MIPI-DSI), we cannot expose 2 devices for the same node (one for the controller, one for the backlight). Some devices, expose backlight as a completely independent controller, see e.g. https://github.com/zephyrproject-rtos/zephyr/blob/23fb3b77a2b2cf53eff1697888f443236069b30c/drivers/misc/grove_lcd_rgb/grove_lcd_rgb.c#L19. This is now poorly handled, and the best you can do it to overcomplicate things by creating a LED driver (API is quite awful), and add overhead by partially wrapping it through the display API. I think we need to solve this properly and not add more ad-hoc solutions.

**Comment by danieldegrasse** [comment]
> I guess the workaround (or actually correct?) would be to expose them as child-nodes? Do we have any example from Linux to compare against?

It does not look like MIPI displays that support setting their backlight via DCS commands use a separate DT node in Linux- they just register a second device object with backlight operations:

https://github.com/torvalds/linux/blob/bc83b4d1f08695e85e85d36f7b803da58010161d/drivers/gpu/drm/panel/panel-startek-kd070fhfid015.c#L273

I suppose we could define backlight devices with `DEVICE_DEFINE`? I'm not sure if I like that approach though, because then it is not clear to users how they access the backlight device. It seems like we might need child nodes here, but that would deviate from Linux. 

Display drivers in Linux register a backlight driver for a given panel with the DRM subsystem via [devm_backlight_device_register](https://github.com/torvalds/linux/blob/bc83b4d1f08695e85e85d36f7b803da58010161d/drivers/video/backlight/backlight.c#L620). This function allocates a new device object for the backlight, and exposes properties of the backlight via sysfs. MIPI displays using DCS would call this function directly and implement support for setting brightness within the same driver, while displays using an external backlight would call [drm_panel_of_backlight](https://github.com/torvalds/linux/blob/bc83b4d1f08695e85e85d36f7b803da58010161d/drivers/gpu/drm/panel/panel-simple.c#L699) to probe and setup their backlight, which would be referenced via a phandle property named "backlight".

To summarize, I think we probably would have to deviate from Linux here- the issue I see is that we don't access devices via sysfs, we access them via devicetree. That means for the MIPI display case we will need a second DT node (probably a child node) that registers as a backlight device.

**Comment by fabiobaltieri** [comment]
@MrMarteng so you want to give a stab and making this a whole subsystem? Seems like where's the suggestions are pointing to.

**Comment by gmarull** [comment]
> > I guess the workaround (or actually correct?) would be to expose them as child-nodes? Do we have any example from Linux to compare against?
> 
> It does not look like MIPI displays that support setting their backlight via DCS commands use a separate DT node in Linux- they just register a second device object with backlight operations:
> 
> https://github.com/torvalds/linux/blob/bc83b4d1f08695e85e85d36f7b803da58010161d/drivers/gpu/drm/panel/panel-startek-kd070fhfid015.c#L273
> 
> I suppose we could define backlight devices with `DEVICE_DEFINE`? I'm not sure if I like that approach though, because then it is not clear to users how they access the backlight device. It seems like we might need child nodes here, but that would deviate from Linux.
> 
> Display drivers in Linux register a backlight driver for a given panel with the DRM subsystem via [devm_backlight_device_register](https://github.com/torvalds/linux/blob/bc83b4d1f08695e85e85d36f7b803da58010161d/drivers/video/backlight/backlight.c#L620). This function allocates a new device object for the backlight, and exposes properties of the backlight via sysfs. MIPI displays using DCS would call this function directly and implement support for setting brightness within the same driver, while displays using an external backlight would call [drm_panel_of_backlight](https://github.com/torvalds/linux/blob/bc83b4d1f08695e85e85d36f7b803da58010161d/drivers/gpu/drm/panel/panel-simple.c#L699) to probe and setup their backlight, which would be referenced via a phandle property named "backlight".
> 
> To summarize, I think we probably would have to deviate from Linux here- the issue I see is that we don't access devices via sysfs, we access them via devicetree. That means for the MIPI display case we will need a second DT node (probably a child node) that registers as a backlight device.

In theory DT nodes should only be created in you need specific properties, ie, backlight has some specific stuff unrelated to the display. Since we have the 1:1 limitation of DT nodes/devices, the only way to escape in our case is to create a child node that represents the backlight.

In the future, we should likely rethink the device model entirely. We should be able to create >1 device per DT node (at least 1 per class), create devices dynamically, devices decoupled from DT (e.g. all the input/lvgl stuff), strongly typed devices, etc., etc. The current model is weak and too rigid. But for now, we can work around the limitations by introducing some DT antipatterns.

File: dts/bindings/display/backlight-pwm.yaml (hunk 1 above)
Comment (decsny) [inline dts/bindings/display/backlight-pwm.yaml:6]: should be compatible `pwm-backlight` to align with linux

File: dts/bindings/display/backlight.yaml (hunk 4 above)
Comment (decsny) [reply dts/bindings/display/backlight.yaml:19]: yes, it has to be redefined, for two reasons. 
1) label is deprecated in base.yaml (see #80863)
2) according to DT spec, eaach device binding should describe what does the label mean for that device

File: dts/bindings/display/backlight.yaml [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    type: string
+    description: |
+      Human readable string describing the Backlight. It can be used by an
+      application to identify this Backlight or to retrieve its
+      number/index (i.e. child node number) on the parent device.
```
Comment (decsny) [inline dts/bindings/display/backlight.yaml:24]: this does not sound like how a label property should be used, what do you mean by this ? in the PR I don't see any information about backlight_info struct either which contains this string ?

File: dts/bindings/display/backlight.yaml [hunk 11]
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2024, Martin Kiepfer
+# SPDX-License-Identifier: Apache-2.0
+
+include: [base.yaml]
+
+properties:
+  init-level:
```
Comment (decsny) [inline dts/bindings/display/backlight.yaml:7]: should this be `default-brightness` to match linux or is it different?

File: dts/bindings/display/backlight.yaml [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    type: int
+    default: 80
```
Comment (decsny) [inline dts/bindings/display/backlight.yaml:9]: default value must have justification in description https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values

File: dts/bindings/display/backlight.yaml [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    description: Initial value of the Backlight.
```
Comment (decsny) [inline dts/bindings/display/backlight.yaml:10]: not clear what the value means from the DT description (I can see it becomes the pulse in code, but user should not have to look for that)

File: dts/bindings/display/backlight.yaml [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+  disable-init-on-boot:
+    type: boolean
+    description: |
+      Disable the initial value during bootup of the device. By default the
+      init-level is applied on boot. This option can prevent from this default
+      behaviour. Backlight needs to be enabled manually later in this case.
```
Comment (decsny) [inline dts/bindings/display/backlight.yaml:17]: not sure also what is the reason for having this property, what would the effect of "not having an initial value" be anyways?  can 0 just mean that for the other property? this property is also not used in the PR?

File: dts/bindings/display/backlight-pwm.yaml [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    type: phandle-array
+    description: |
+      Reference to a PWM instance.
+
+      The period field is used by the set_brightness function of the
+      Backlight-PWM API.
+      Its value should at least be greater than 100 nanoseconds (for a full
+      brigtness granularity) and lesser than 50 milliseconds (average visual
+      persistence time of the human eye). Typical values for the PWM period
+      are 10 or 20 milliseconds.
```
Comment (decsny) [inline dts/bindings/display/backlight-pwm.yaml:22]: this is just copy pasted from [pwm-leds.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/4300a4c33b0f6becc43d1e3a130ebee19c411c6b/dts/bindings/led/pwm-leds.yaml#L17-L21) ? is it accurate still ? Also, IMO the description in that binding is weak, see other examples in other bindings where they give an example / format of the array specifier

File: dts/bindings/display/backlight-pwm.yaml (hunk 1 above)
Comment (decsny) [reply dts/bindings/display/backlight-pwm.yaml:6]: no, we should not do that here, there is a linux compatible we should match

**Review by decsny — CHANGES_REQUESTED** [review]

**Comment by decsny** [comment]
> In the future, we should likely rethink the device model entirely. We should be able to create >1 device per DT node (at least 1 per class), create devices dynamically, devices decoupled from DT (e.g. all the input/lvgl stuff), strongly typed devices, etc., etc. The current model is weak and too rigid. But for now, we can work around the limitations by introducing some DT antipatterns.

technically, you can make multiple device per DT node, it's just you can only make one device per DT node using the DEVICE_DT_DEFINE macro, because there can only be one ordinal number per node, so you would have to do it manually. you could maybe have like a parent driver that gets the ordinal node and makes all the devices manually and knows the order to initialize them ? IDK though because then I guess you lose some flexibility on init dependencies with other things, it's just a hacked idea
